### PR TITLE
Add rnd_ppo.py documentation and refactor

### LINKED ISSUE #127: Add `rnd_ppo.py` documentation and refactor
`rnd_ppo.py` is a bit dated, and I recommend refactoring it to match other PPO style, which would include:

- [x] change the name from `rnd_ppo.py` to `ppo_rnd.py`
- [x] use `from gym.wrappers.normalize import RunningMeanStd` instead of the implementing ourselves (note the implementation might be a bit different).
- [x] create a `make_env` function like https://github.com/vwxyzjn/cleanrl/blob/0b3f8eae7d07b90a0ee129ffe290bd82e5b57a14/cleanrl/ppo_atari.py#L88-L103
- [x] remove the visualization (i.e., `ProbsVisualizationWrapper`)
- [x] use ` def get_value` and `def get_action_and_value` for the `Agent` class
- [x] remove https://github.com/vwxyzjn/cleanrl/blob/0b3f8eae7d07b90a0ee129ffe290bd82e5b57a14/cleanrl/rnd_ppo.py#L706-L708
- [x] maybe log the average `curiosity_reward` instead? https://github.com/vwxyzjn/cleanrl/blob/0b3f8eae7d07b90a0ee129ffe290bd82e5b57a14/cleanrl/rnd_ppo.py#L848
- [x] name `total_reward_per_env` to `curiosity_return` https://github.com/vwxyzjn/cleanrl/blob/0b3f8eae7d07b90a0ee129ffe290bd82e5b57a14/cleanrl/rnd_ppo.py#L854
- [x] Add SPS (steps per second) metric. 

Overall I suggest selecting `ppo_atari.py` and `rnd_ppo.py` and use `Compare Selected` on VSCode to see the file difference and minimize the file difference:

image

## Types of changes

- [ ] Bug fix
- [ ] New feature
- [ ] New algorithm
- [x] Documentation

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. 
- [x] I have contacted @vwxyzjn to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
    - [x] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [x] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [ ] I have added links to the tracked experiments.
- [ ] I have updated the tests accordingly (if applicable).

### PR DESCRIPTION
**Pull request by yooceii**
## Description

Closes #127 

## Types of changes

- [ ] Bug fix
- [ ] New feature
- [x] New algorithm
- [x] Documentation

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [x] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [x] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [x] I have contacted @vwxyzjn to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [x] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [x] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [x] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [ ] I have added links to the tracked experiments.
- [x] I have updated the tests accordingly (if applicable).

### CODE REVIEW

**Comment by vwxyzjn** [comment]
Hey, @yooceii would you mind reverting the formatting change? They make it harder to review and identify the code specifically relating to RND. 

image

Formatting change should be done in a larger PR in #167 together...

File: benchmark/rnd.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# export WANDB_ENTITY=openrlbenchmark
+
+poetry install -E envpool
+xvfb-run -a python -m cleanrl_utils.benchmark \
+    --env-ids MontezumaRevenge-v5 \
+    --command "poetry run python cleanrl/ppo_atari_envpool.py --total-timesteps 2000000000 --ent-coef 0.001 --num-envs 128 --gamma 0.999 --int-gamma 0.99 --learning-rate 0.0001 --track --capture-video" \
+    --num-seeds 3 \
```
Comment (vwxyzjn) [inline benchmark/rnd.sh:7]: If we only used 1 random seed it should be 1 here, too.  The command was wrong, it should be `ppo_rnd_envpool.py`

File: cleanrl/ppo_rnd_envpool.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+        help="if toggled, cuda will be enabled by default")
+    parser.add_argument("--track", type=lambda x: bool(strtobool(x)), default=False, nargs="?", const=True,
+        help="if toggled, this experiment will be tracked with Weights and Biases")
+    parser.add_argument("--wandb-project-name", type=str, default="cleanRL",
+        help="the wandb's project name")
+    parser.add_argument("--wandb-entity", type=str, default=None,
+        help="the entity (team) of wandb's project")
+
+    # Algorithm specific arguments
+    parser.add_argument("--env-id", type=str, default="BreakoutNoFrameskip-v4",
```
Comment (vwxyzjn) [inline cleanrl/ppo_rnd_envpool.py:38]: Breakout? Should this be `MontezumaRevenge-v5`?

File: cleanrl/ppo_rnd_envpool.py [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+
+    # env setup
+    envs = envpool.make(
+        args.env_id,
+        env_type="gym",
+        num_envs=args.num_envs,
+        episodic_life=True,
+        reward_clip=True,
+        seed=args.seed,
+        repeat_action_probability=0.25,
```
Comment (vwxyzjn) [inline cleanrl/ppo_rnd_envpool.py:304]: It's worth mentioning we are using this sticky action `repeat_action_probability=0.25` in the implementation details section in the docs (i.e., `ppo-rnd.md`).

File: cleanrl/ppo_rnd_envpool.py [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    start_time = time.time()
+    next_obs = torch.Tensor(envs.reset()).to(device)
+    next_done = torch.zeros(args.num_envs).to(device)
+    num_updates = args.total_timesteps // args.batch_size
+
+    print("Start to initialize observation normalization parameter.....")
+    next_ob = []
+    for step in range(args.num_steps * 50):
```
Comment (vwxyzjn) [inline cleanrl/ppo_rnd_envpool.py:344]: Where is this `50` coming from? The only thing I found in the paper is " We initialize the normalization parameters by stepping a random agent in the environment for a small number of steps before beginning optimization."

We should also add this in the implementation details section in the docs (i.e., ppo-rnd.md).

File: cleanrl/ppo_rnd_envpool.py [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+            np.std(curiosity_reward_per_env),
+            len(curiosity_reward_per_env),
+        )
+        reward_rms.update_from_moments(mean, std**2, count)
+
+        curiosity_rewards /= np.sqrt(reward_rms.var)
+
+        # bootstrap reward if not done. reached the batch limit
+        with torch.no_grad():
+            last_value_ext, last_value_int = agent.get_value(next_obs)
```
Comment (vwxyzjn) [inline cleanrl/ppo_rnd_envpool.py:422]: Should use `next_value_ext` to be consistent with other PPO implementations.

File: cleanrl/ppo_rnd_envpool.py [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+        b_ext_returns = ext_returns.reshape(-1)
+        b_int_returns = int_returns.reshape(-1)
+        b_ext_values = ext_values.reshape(-1)
+
+        b_advantages = b_int_advantages * args.int_coef + b_ext_advantages * args.ext_coef
+
+        obs_rms.update(b_obs[:, 3, :, :].reshape(-1, 1, 84, 84).cpu().numpy())
+
+        # Optimizing the policy and value network
+        forward_mse = nn.MSELoss(reduction="none")
```
Comment (vwxyzjn) [inline cleanrl/ppo_rnd_envpool.py:484]: Could we use `F.mse_loss(reduction="none")` to be consistent with DDPG and TD3? Alternatively, could we just do `(x - y) **2`?

See 
https://github.com/vwxyzjn/cleanrl/blob/1878a4f5e29e848135c6e08f89afdaa744403b48/cleanrl/td3_continuous_action.py#L223-L224

**Review by vwxyzjn — CHANGES_REQUESTED** [review]
Looks good! Left a few comments.

**Review by vwxyzjn — CHANGES_REQUESTED** [review]
Looks great! Have a couple of comments.

File: cleanrl/ppo_rnd_envpool.py (hunk 4 above)
Comment (yooceii) [reply cleanrl/ppo_rnd_envpool.py:344]: It's from https://github.com/openai/random-network-distillation/blob/f75c0f1efa473d5109d487062fd8ed49ddce6634/run_atari.py#L69

**Comment by vwxyzjn** [comment]
## Refactor Check (compatible with the performance in the tracked experiment)


I compared the SPS performance of the latest refactor against the [old script used in the tracked experiment](https://wandb.ai/openrlbenchmark/cleanrl/runs/1wm98fjm/code?workspace=user-costa-huang). I named the old one `old_ppo_rnd_envpool.py` and ran it and the latest script without initializing observation normalization parameter.

The following screenshot confirms the refactor does not result in a performance difference (and that we did the refactor correctly — making it faster without impacting sample efficiency)

image

**Review by vwxyzjn — CHANGES_REQUESTED** [review]
Thank you @yooceii! There is one last thing. Could you also document the meaning of `losses/fwd_loss` in the docs? Otherwise, everything LGTM. The comment above also confirms the refactor was only beneficial (reducing GPU memory, making the script about 50% faster)

**Review by vwxyzjn — APPROVED** [review]
Thanks for the great work @yooceii!
